exist-core/src/main/java/org/exist/xquery/XQueryContext.java

```diff
@@ -3322,39 +3185,36 @@ public boolean requireDebugMode() {
     private Deque<BinaryValue> binaryValueInstances;
 
     void enterEnclosedExpr() {
-        if (binaryValueInstances != null) {
-            final Iterator<BinaryValue> it = binaryValueInstances.descendingIterator();
-            while (it.hasNext()) {
-                it.next().incrementSharedReferences();
-            }
+        if (binaryValueInstances == null) {
+            return;
+        }
+        final Iterator<BinaryValue> it = binaryValueInstances.descendingIterator();
+        while (it.hasNext()) {
+            it.next().incrementSharedReferences();
         }
     }
 
     void exitEnclosedExpr() {
-        if (binaryValueInstances != null) {
-            final Iterator<BinaryValue> it = binaryValueInstances.iterator();
-            List<BinaryValue> destroyable = null;
-            while (it.hasNext()) {
-                try {
-                    final BinaryValue bv = it.next();
-                    bv.close(); // really just decrements a reference
-                    if (bv.isClosed()) {
-                        if (destroyable == null) {
-                            destroyable = new ArrayList<>();
-                        }
-                        destroyable.add(bv);
-                    }
-                } catch (final IOException e) {
-                    LOG.warn("Unable to close binary reference on exiting enclosed expression: {}", e.getMessage(), e);
+        if (binaryValueInstances == null) {
+            return;
+        }
+        final Iterator<BinaryValue> it = binaryValueInstances.iterator();
+        final List<BinaryValue> destroyable = new ArrayList<>();
```

Comment:
Another issue with a change to eager evaluation from lazy evaluation. If there are no binary value instances which is likely true for many many XQuery's then this function will use more memory and be less performant. As this is called for every enclosed expression, not wasting memory and performance here is critical.